uji jarque bera

In [ ]:
from scipy.stats import chi2

def uji_jarque_bera(data, alpha=0.05):
    # 1. Hitung ukuran data dan rata-rata
    n = len(data)
    mean = sum(data) / n

    # 2. Hitung selisih tiap data dengan rata-rata
    selisih = [x - mean for x in data]

    # 3. Hitung Momen (m2, m3, m4)
    m2 = sum(d**2 for d in selisih) / n
    m3 = sum(d**3 for d in selisih) / n  # Sudah diperbaiki menjadi pangkat 3
    m4 = sum(d**4 for d in selisih) / n

    # 4. Hitung Skewness dan Kurtosis
    skewness = m3 / (m2**1.5)
    kurtosis = m4 / (m2**2)

    # 5. Hitung Statistik Hitung Jarque-Bera (JB)
    chi_square = (n / 6) * (skewness**2 + ((kurtosis - 3)**2) / 4)

    # 6. Tentukan Nilai Kritis Tabel (df selalu 2 untuk Jarque-Bera)
    df = 2
    chi_tabel = chi2.ppf(1 - alpha, df)

    # 7. Aturan Keputusan
    if chi_square > chi_tabel:
        keputusan = "Tolak H0: data tidak berdistribusi normal"
    else:
        keputusan = "Gagal menolak H0: data berdistribusi normal"

    # Mengembalikan seluruh nilai yang diminta soal
    return n, skewness, kurtosis, chi_square, chi_tabel, keputusan

# --- CARA MENGGUNAKAN FUNGSINYA ---
data_sampel = [12, 15, 14, 10, 18, 16, 13, 17, 11, 14]
n, skew, kurt, jb_hitung, jb_tabel, hasil = uji_jarque_bera(data_sampel)

print(f"Banyak Data (n) : {n}")
print(f"Skewness        : {skew:.4f}")
print(f"Kurtosis        : {kurt:.4f}")
print(f"JB Hitung       : {jb_hitung:.4f}")
print(f"JB Tabel        : {jb_tabel:.4f}")
print(f"Keputusan       : {hasil}")

Banyak Data (n) : 10
Skewness        : 0.0000
Kurtosis        : 1.9667
JB Hitung       : 0.4449
JB Tabel        : 5.9915
Keputusan       : Gagal menolak H0: data berdistribusi normal


Output ini menunjukkan bahwa dari 10 data yang diuji, bentuk sebarannya terbukti sangat simetris seimbang antara sisi kanan dan kiri (Skewness = 0.0000) meskipun sedikit lebih landai dibandingkan bentuk lonceng normal ideal (Kurtosis = 1.9667). Nilai JB Hitung sebesar 0.4449 merupakan skor total penyimpangan atau "keanehan" bentuk data kita; karena skor ini jauh lebih kecil daripada batas toleransi maksimal yang diizinkan oleh JB Tabel (5.9915), maka ketidaknormalan data tersebut dianggap sangat kecil dan tidak berarti. Oleh karena itu, keputusannya adalah Gagal menolak H0, yang artinya data kalian secara statistik terbukti sah berdistribusi normal dan sudah aman serta valid untuk digunakan pada tahap analisis atau pemodelan sains data selanjutnya.

uji kolmogorov smirnov

In [ ]:
from scipy.stats import norm, kstwobign

def uji_kolmogorov_smirnov(data, alpha=0.05):
    n = len(data)
    # 1. Urutkan data wajib hukumnya!
    data_urut = sorted(data)

    # 2. Hitung Mean & Standar Deviasi
    mean = sum(data_urut) / n
    varians = sum((x - mean)**2 for x in data_urut) / (n - 1)
    std_dev = varians ** 0.5

    # 3. Cari Selisih Terbesar (D)
    max_d = 0
    for i in range(n):
        x = data_urut[i]

        # Nilai Z dan Kumulatif Teoretis (F0)
        z = (x - mean) / std_dev
        f0 = norm.cdf(z)

        # Kumulatif Empiris (Sn)
        sn = (i + 1) / n

        # Hitung selisih mutlaknya
        selisih = abs(f0 - sn)

        # Simpan jika ini selisih terbesar
        if selisih > max_d:
            max_d = selisih

    # 4. Tentukan D Tabel (Pendekatan Asimtotik untuk KS)
    # Rumus umum untuk alpha = 0.05 dan n > 35 biasanya adalah 1.36 / sqrt(n)
    d_tabel = 1.36 / (n ** 0.5)

    # 5. Aturan Keputusan
    if max_d > d_tabel:
        keputusan = "Tolak H0: data tidak berdistribusi normal"
    else:
        keputusan = "Gagal menolak H0: data berdistribusi normal"

    return max_d, d_tabel, keputusan

# --- CARA MENGGUNAKAN FUNGSINYA ---
data_sampel = [12, 15, 14, 10, 18, 16, 13, 17, 11, 14]
d_hitung, d_tabel, hasil_ks = uji_kolmogorov_smirnov(data_sampel)

print(f"D Hitung  : {d_hitung:.4f}")
print(f"D Tabel   : {d_tabel:.4f}")
print(f"Keputusan : {hasil_ks}")

D Hitung  : 0.1000
D Tabel   : 0.4301
Keputusan : Gagal menolak H0: data berdistribusi normal


Rumus
Standarisasi Nilai Z (Z-Score):$$Z_i = \frac{x_i - \bar{x}}{s}$$Fungsi Distribusi Kumulatif Empiris ($S_n(x_i)$):$$S_n(x_i) = \frac{i}{n}$$Fungsi Distribusi Kumulatif Teoretis ($F_0(x_i)$):Diperoleh dari nilai peluang kurva normal standar $P(Z \le Z_i)$.Statistik Uji Kolmogorov-Smirnov ($D$):$$D = \max |F_0(x_i) - S_n(x_i)|$$

Langkah-Langkah PemrogramanUrutkan data sampel dari nilai terkecil hingga terbesar (wajib).Hitung nilai rata-rata (mean) dan standar deviasi sampel.Cari nilai Z-score untuk setiap data, lalu hitung peluang kumulatif teoretisnya ($F_0$).Hitung peluang kumulatif empirisnya ($S_n$) dengan rumus indeks data dibagi total data ($i/n$).Cari selisih mutlak (positif) antara $F_0$ dan $S_n$ pada setiap baris data.Ambil nilai selisih yang paling besar sebagai nilai statistik $D_{hitung}$.Bandingkan nilai $D_{hitung}$ dengan batas toleransi dari $D_{tabel}$.

uji F varians

In [ ]:
from scipy.stats import f

def uji_f_varians(data1, data2, alpha=0.05):
    # 1. Menghitung banyak data (n1 dan n2)
    n1 = len(data1)
    n2 = len(data2)

    # 2. Menghitung rata-rata (mean) kedua kelompok
    mean1 = sum(data1) / n1
    mean2 = sum(data2) / n2

    # 3. Menghitung varians sampel kelompok 1 (dibagi n-1)
    varians1 = sum((x - mean1)**2 for x in data1) / (n1 - 1)

    # 4. Menghitung varians sampel kelompok 2 (dibagi n-1)
    varians2 = sum((x - mean2)**2 for x in data2) / (n2 - 1)

    # 5. Menentukan Pembilang & Penyebut untuk F Hitung (Varians Besar / Varians Kecil)
    if varians1 > varians2:
        f_hitung = varians1 / varians2
        df1 = n1 - 1
        df2 = n2 - 1
    else:
        f_hitung = varians2 / varians1
        df1 = n2 - 1
        df2 = n1 - 1

    # 6. Menghitung nilai F Tabel secara otomatis
    f_tabel = f.ppf(1 - alpha, df1, df2)

    # 7. Aturan Keputusan
    if f_hitung > f_tabel:
        keputusan = "Tolak H0: varians tidak homogen"
    else:
        keputusan = "Gagal menolak H0: varians homogen"

    # Mengembalian seluruh nilai sesuai permintaan soal
    return n1, n2, varians1, varians2, f_hitung, f_tabel, keputusan

# --- CONTOH MENJALANKAN KODE ---
kelas_A = [70, 75, 80, 72, 78]
kelas_B = [65, 85, 60, 90, 70]

n1, n2, var1, var2, f_hit, f_tab, hasil = uji_f_varians(kelas_A, kelas_B)
print(f"n1: {n1}, n2: {n2}")
print(f"Varians 1: {var1:.4f}, Varians 2: {var2:.4f}")
print(f"F Hitung : {f_hit:.4f} | F Tabel : {f_tab:.4f}")
print(f"Keputusan: {hasil}")

n1: 5, n2: 5
Varians 1: 17.0000, Varians 2: 167.5000
F Hitung : 9.8529 | F Tabel : 6.3882
Keputusan: Tolak H0: varians tidak homogen


Output ini menunjukkan terdapat perbedaan tingkat penyebaran data yang sangat jomplang antara kelompok pertama (Varians = 17.0) yang datanya cenderung mengumpul rapat, dengan kelompok kedua (Varians = 167.5) yang datanya sangat menyebar luas. Nilai F Hitung sebesar 9.8529 merupakan skor penanda ketidakseimbangan sebaran tersebut; karena skor keanehan ini ternyata lebih besar daripada batas toleransi maksimal yang diizinkan oleh F Tabel (6.3882), maka perbedaan tersebut dinilai terlalu besar dan tidak bisa dianggap sebagai faktor kebetulan belaka. Oleh karena itu, keputusannya adalah Tolak H0, yang artinya secara statistik terbukti sah bahwa varians kedua kelompok data tersebut tidak homogen (tidak setara) sehingga belum memenuhi syarat ideal untuk langsung dibandingkan.

uji homogenitas

In [ ]:
from scipy.stats import f

def uji_levene(data1, data2, alpha=0.05):
    n1, n2 = len(data1), len(data2)
    N = n1 + n2

    # 1. Hitung rata-rata data asli kelompok
    m1 = sum(data1) / n1
    m2 = sum(data2) / n2

    # 2. Langkah Transformasi Levene: Cari Nilai Absolut Deviasi (Z)
    z1 = [abs(x - m1) for x in data1]
    z2 = [abs(x - m2) for x in data2]

    # 3. Hitung rata-rata dari nilai Z
    mz1 = sum(z1) / n1
    mz2 = sum(z2) / n2
    grand_mz = (sum(z1) + sum(z2)) / N

    # 4. Hitung Jumlah Kuadrat Antar Kelompok (SSB) dan Dalam Kelompok (SSW)
    ssb = n1 * (mz1 - grand_mz)**2 + n2 * (mz2 - grand_mz)**2
    ssw = sum((z - mz1)**2 for z in z1) + sum((z - mz2)**2 for z in z2)

    # 5. Hitung Mean Squares (Rata-rata Jumlah Kuadrat)
    msb = ssb / (2 - 1)    # df antara kelompok = k - 1 = 2 - 1 = 1
    msw = ssw / (N - 2)    # df dalam kelompok = N - k = N - 2

    # 6. Menghitung Statistik Uji Levene (W Hitung)
    w_hitung = msb / msw
    w_tabel = f.ppf(1 - alpha, 1, N - 2)

    # 7. Aturan Keputusan
    if w_hitung > w_tabel:
        keputusan = "Tolak H0: varians tidak homogen"
    else:
        keputusan = "Gagal menolak H0: varians homogen"

    return w_hitung, w_tabel, keputusan

# --- CONTOH MENJALANKAN KODE ---
kelas_A = [70, 75, 80, 72, 78]
kelas_B = [65, 85, 60, 90, 70]

w_hit, w_tab, hasil_levene = uji_levene(kelas_A, kelas_B)
print(f"W Hitung Levene: {w_hit:.4f} | W Tabel: {w_tab:.4f}")
print(f"Keputusan      : {hasil_levene}")

W Hitung Levene: 11.1506 | W Tabel: 5.3177
Keputusan      : Tolak H0: varians tidak homogen


Rumus Transformasi Data ($Z_{ij}$):$$Z_{ij} = |Y_{ij} - \bar{Y}_i|$$di mana $Y_{ij}$ adalah data asli ke-$j$ pada kelompok ke-$i$, dan $\bar{Y}_i$ adalah rata-rata kelompok ke-$i$.Menghitung Rata-rata Nilai $Z$ Kelompok ($\bar{Z}_{i\cdot}$) dan Rata-rata Gabungan ($\bar{Z}_{\cdot\cdot}$):$$\bar{Z}_{1\cdot} = \frac{\sum Z_{1j}}{n_1}, \quad \bar{Z}_{2\cdot} = \frac{\sum Z_{2j}}{n_2}, \quad \bar{Z}_{\cdot\cdot} = \frac{\sum Z_{ij}}{N}$$Statistik Uji Levene ($W$ atau F-Levene):$$W = \frac{(N - 2)}{(2 - 1)} \cdot \frac{n_1(\bar{Z}_{1\cdot} - \bar{Z}_{\cdot\cdot})^2 + n_2(\bar{Z}_{2\cdot} - \bar{Z}_{\cdot\cdot})^2}{\sum(Z_{1j} - \bar{Z}_{1\cdot})^2 + \sum(Z_{2j} - \bar{Z}_{2\cdot})^2}$$

Langkah-Langkah Pemrograman Hitung rata-rata data asli untuk masing-masing kelompok. Buat data baru ($Z$) yang isinya jarak/selisih positif (absolut) antara data asli dengan rata-ratanya masing-masing.Hitung rata-rata nilai $Z$ kelompok 1, kelompok 2, serta rata-rata total gabungan seluruh nilai $Z$.Hitung nilai kuadrat deviasi antar kelompok (SSB) dan di dalam kelompok (SSW).Bagi masing-masing dengan derajat bebasnya untuk mendapatkan nilai MSB dan MSW.Nilai W Hitung didapatkan dari hasil pembagian MSB / MSW.Bandingkan W Hitung dengan F Tabel distribusi F ($df_1 = 1$, $df_2 = N - 2$).

uji chi square

In [ ]:
from scipy.stats import chi2

def uji_chi_square(data, alpha=0.05):
    # 1. Menghitung jumlah baris dan kolom
    jumlah_baris = len(data)
    jumlah_kolom = len(data[0])

    # 2. Menghitung total data di setiap baris dan kolom
    total_baris = [sum(baris) for baris in data]
    total_kolom = [sum(data[i][j] for i in range(jumlah_baris)) for j in range(jumlah_kolom)]

    # 3. Menghitung grand_total keseluruhan data
    grand_total = sum(total_baris)

    # 4. Menghitung Frekuensi Harapan (harapan)
    harapan = []
    for i in range(jumlah_baris):
        baris_harapan = []
        for j in range(jumlah_kolom):
            eij = (total_baris[i] * total_kolom[j]) / grand_total
            baris_harapan.append(eij)
        harapan.append(baris_harapan)

    # 5. Menghitung Nilai Statistik Uji Chi-Square Hitung
    chi_square = 0
    for i in range(jumlah_baris):
        for j in range(jumlah_kolom):
            O = data[i][j]
            E = harapan[i][j]
            chi_square += ((O - E)**2) / E

    # 6. Menghitung derajat bebas (df) dan Chi-Square Tabel
    df = (jumlah_baris - 1) * (jumlah_kolom - 1)
    chi_tabel = chi2.ppf(1 - alpha, df)

    # 7. Aturan Keputusan
    if chi_square > chi_tabel:
        keputusan = "Tolak H0: terdapat hubungan"
    else:
        keputusan = "Gagal menolak H0: tidak terdapat hubungan"

    # Mengembalikan 7 nilai secara lengkap sesuai permintaan soal
    return jumlah_baris, jumlah_kolom, grand_total, harapan, chi_square, chi_tabel, keputusan

# --- CONTOH MENJALANKAN KODE ---
# Data preferensi Genre Film (Horor/Komedi) berdasarkan Gender (Pria/Wanita)
tabel_kontingensi = [[30, 20],
                     [15, 35]]

j_baris, j_kolom, g_total, mat_ef, chi_hit, chi_tab, hasil = uji_chi_square(tabel_kontingensi)

print(f"Baris: {j_baris} | Kolom: {j_kolom} | Grand Total: {g_total}")
print(f"Chi-Square Hitung: {chi_hit:.4f} | Chi-Square Tabel: {chi_tab:.4f}")
print(f"Keputusan: {hasil}")

Baris: 2 | Kolom: 2 | Grand Total: 100
Chi-Square Hitung: 9.0909 | Chi-Square Tabel: 3.8415
Keputusan: Tolak H0: terdapat hubungan


Output ini menunjukkan dari total 100 data yang diamati dalam tabel kontingensi 2×2, kedua variabel yang kalian uji terbukti saling memengaruhi dan memiliki keterikatan yang nyata. Nilai Chi-Square Hitung sebesar 9.0909 merupakan skor penanda besarnya jurang perbedaan antara kenyataan yang ditemukan di lapangan dengan kondisi ideal teoretisnya. Karena skor perbedaan tersebut ternyata jauh lebih besar daripada batas toleransi maksimal yang diizinkan oleh Chi-Square Tabel (3.8415), maka pola data ini dinilai sangat meyakinkan dan tidak mungkin terjadi hanya karena faktor kebetulan semata. Oleh karena itu, keputusannya adalah Tolak H0, yang artinya secara statistik terbukti sah bahwa terdapat hubungan yang signifikan antara kedua variabel tersebut dalam analisis sains data kita.

uji fisher exac test

In [ ]:
# 1. Membuat fungsi hitung faktorial manual tanpa library
def faktorial(angka):
    hasil = 1
    for i in range(1, angka + 1):
        hasil *= i
    return hasil

def uji_fisher_exact(data, alpha=0.05):
    # 2. Ekstrak data dari matriks 2x2
    a = data[0][0]
    b = data[0][1]
    c = data[1][0]
    d = data[1][1]

    # 3. Hitung jumlahan baris, kolom, dan total n
    r1 = a + b
    r2 = c + d
    c1 = a + c
    c2 = b + d
    n = a + b + c + d

    # 4. Hitung p-value menggunakan rumus Fisher
    pembilang = faktorial(r1) * faktorial(r2) * faktorial(c1) * faktorial(c2)
    penyebut = faktorial(a) * faktorial(b) * faktorial(c) * faktorial(d) * faktorial(n)
    p_value = pembilang / penyebut

    # 5. Aturan Keputusan
    if p_value < alpha:
        keputusan = "Tolak H0: terdapat hubungan signifikan (sampel kecil)"
    else:
        keputusan = "Gagal menolak H0: tidak terdapat hubungan signifikan"

    return p_value, keputusan

# --- CONTOH MENJALANKAN KODE (Sampel Kecil) ---
data_medis = [[1, 9],
              [11, 3]]

p_val, hasil_fisher = uji_fisher_exact(data_medis)
print(f"P-Value Fisher: {p_val:.6f}")
print(f"Keputusan     : {hasil_fisher}")

P-Value Fisher: 0.001346
Keputusan     : Tolak H0: terdapat hubungan signifikan (sampel kecil)


Rumus peluang (p-value) hipergeometris untuk mendapatkan susunan angka tersebut secara eksak adalah:$$p = \frac{\binom{a+b}{a} \binom{c+d}{c}}{\binom{n}{a+c}} = \frac{(a+b)! (c+d)! (a+c)! (b+d)!}{a! b! c! d! n!}$$

Langkah-Langkah Pemrograman Ambil nilai dari keempat sel tabel $2 \times 2$ (tandai sebagai variabel $a, b, c, d$). Hitung total jumlahan untuk masing-masing Baris 1, Baris 2, Kolom 1, Kolom 2, serta Grand Total ($n$). Buat sebuah fungsi pembantu (helper function) untuk menghitung nilai perkalian faktorial secara manual. Masukkan seluruh nilai faktorial dari jumlahan baris, kolom, dan sel ke dalam rumus Fisher's Exact. Hitung nilai probabilitas akhir (p-value).Bandingkan p-value dengan tingkat kekritisan $\alpha = 0.05$. Jika p-value $\le 0.05$, maka Tolak $H_0$.